### PR TITLE
gh-124153: Implement `PyType_GetBaseByToken()` and `Py_tp_token` slot

### PR DESCRIPTION
**Pull request by neonene**
* Issue: gh-124153

----
📚 Documentation preview 📚: https://cpython-previews--124163.org.readthedocs.build/

### CODE REVIEW

**Comment by neonene** [comment]
I'll check this and PoC version again.

File: Objects/typeobject.c [hunk 1]
```diff
@@ -5274,6 +5287,129 @@ _PyType_GetModuleByDef2(PyTypeObject *left, PyTypeObject *right,
...
+}
+
+int
+PyType_GetBaseByToken(PyTypeObject *type, void *token, PyTypeObject **result)
+{
+#define RETURN_RESULT(BASE, RET) \
+    do { \
+        *result = (PyTypeObject *)(BASE); \
+        return (RET); \
+    } while (0)
```
Comment (encukou) [inline Objects/typeobject.c]: This macro adds 6 lines to avoid 5 simpler lines. Consider writing the assignments out instead.

**Review by encukou — APPROVED** [review]
Looks great, thank you!
I have just one style suggestion. But if you'd like to keep it we can merge the PR as it is.

File: Objects/typeobject.c [hunk 2]
```diff
@@ -5274,6 +5287,130 @@ _PyType_GetModuleByDef2(PyTypeObject *left, PyTypeObject *right,
...
+                     "PyType_GetBaseByToken called with token=NULL");
+        return -1;
+    }
+    if (!PyType_Check(type)) {
+        PyErr_Format(PyExc_TypeError,
+                     "expected a type, got a '%T' object", type);
+        return -1;
+    }
+
+    if (!_PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE)) {
```
Comment (neonene) [inline Objects/typeobject.c]: When I observed a good performance, the line had a typo (missing a NOT), though `check_base_by_token()` was unused during a PGO test.  The `telco` got slower by 5-10% once I corrected, which was actually fixed by one of the following:

* Use `if-else` statement

* Replace `_PyType_HasFeature()` with the public `PyType_HasFeature()`, not with ineffective `(type->tp_flags & Py_TPFLAGS_HEAPTYPE)`

They are not used in `PyType_GetBaseByToken()`.

**Comment by neonene** [comment]
I think I've finished.  When checking a type without a result, this may be a bit slower than the draft PR, not trained well for now.

**Pull request merged**